### PR TITLE
Adding ZLIB custom memory management

### PR DESCRIPTION
**Pull request by sdavtaker**
*Issue #, if available:*

*Description of changes:*

*Check all that applies:*
- [x] Did a review by yourself.
- [x] Added proper tests to cover this PR. (If tests are not applicable, explain.)
- [x] Checked if this PR is a breaking (APIs have been changed) change.
- [x] Checked if this PR will _not_ introduce cross-platform inconsistent behavior.
- [x] Checked if this PR would require a ReadMe/Wiki update.

Check which platforms you have built SDK on to verify the correctness of this PR.
- [x] Linux
- [x] Windows
- [ ] Android
- [x] MacOS
- [ ] IOS
- [ ] Other Platforms

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/aws-cpp-sdk-core/source/client/RequestCompression.cpp [hunk 1]
```diff
@@ -37,6 +38,40 @@ Aws::String Aws::Client::GetCompressionAlgorithmId(const CompressionAlgorithm &a
     }
 }
 
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+#ifdef USE_AWS_MEMORY_MANAGEMENT
+static const char* ZlibMemTag = "zlib";
+static size_t offset = sizeof(size_t); // to make space for size of the array
```
Comment (SergeyRyabinin) [inline src/aws-cpp-sdk-core/source/client/RequestCompression.cpp:44]: add const
static variables are not const

File: src/aws-cpp-sdk-core/source/client/RequestCompression.cpp [hunk 2]
```diff
@@ -37,6 +38,40 @@ Aws::String Aws::Client::GetCompressionAlgorithmId(const CompressionAlgorithm &a
...
+//Define custom memory allocation for zlib
+// if fail to allocate, return Z_NULL
+void* aws_zalloc(void * /* opaque */, unsigned items, unsigned size)
+{
+    unsigned sizeToAllocate = items*size;
```
Comment (SergeyRyabinin) [inline src/aws-cpp-sdk-core/source/client/RequestCompression.cpp:49]: check for `unsigned` overflow

File: src/aws-cpp-sdk-core/source/client/RequestCompression.cpp [hunk 3]
```diff
@@ -37,6 +38,40 @@ Aws::String Aws::Client::GetCompressionAlgorithmId(const CompressionAlgorithm &a
...
+    char* newMem = reinterpret_cast<char*>(Aws::Malloc(ZlibMemTag, sizeToAllocate + offset));
```
Comment (SergeyRyabinin) [inline src/aws-cpp-sdk-core/source/client/RequestCompression.cpp:50]: check for `sizeToAllocate + offset` overflow

File: src/aws-cpp-sdk-core/source/client/RequestCompression.cpp [hunk 4]
```diff
@@ -37,6 +38,40 @@ Aws::String Aws::Client::GetCompressionAlgorithmId(const CompressionAlgorithm &a
...
+    }
+    else
+    {
+        return Z_NULL;
+    }
+}
+
+void aws_zfree(void * /* opaque */, void * ptr)
+{
+    if(ptr)
```
Comment (SergeyRyabinin) [inline src/aws-cpp-sdk-core/source/client/RequestCompression.cpp:64]: ptr && ptr > offset

Comment (sdavtaker) [reply src/aws-cpp-sdk-core/source/client/RequestCompression.cpp:70]: that offset can be only 1, 4 or maybe 8. That is a way very little number to be valid memory anyway.

**Review by SergeyRyabinin — APPROVED** [review]

**Pull request merged**